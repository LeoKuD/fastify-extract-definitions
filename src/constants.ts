import { CachedSchemas, Method } from './types';

export const rootName = 'API';

export const bannerComment = `/* tslint:disable */
/**
 * This file was automatically generated by fastify-extract-definitions.
 * DO NOT MODIFY IT BY HAND. Instead, modify the fastify routes schemas
 * and re-run project to regenerate this file.
 */\n
`;

export const bannerImports = `import {
  ContextConfigDefault,
  RawReplyDefaultExpression,
  RawRequestDefaultExpression,
  RawServerDefault,
  RouteHandlerMethod,
} from 'fastify';
import { RouteGenericInterface } from 'fastify/types/route';

export type Handler<
  RouteGeneric extends RouteGenericInterface = RouteGenericInterface,
  ContextConfig = ContextConfigDefault
> = RouteHandlerMethod<
  RawServerDefault,
  RawRequestDefaultExpression,
  RawReplyDefaultExpression,
  RouteGeneric,
  ContextConfig
>;\n
`;

export const cachedMethods: Record<
  Method,
  { capitalized: string; uppercase: string }
> = {
  delete: { capitalized: 'Delete', uppercase: 'DELETE' },
  get: { capitalized: 'Get', uppercase: 'GET' },
  head: { capitalized: 'Head', uppercase: 'HEAD' },
  options: { capitalized: 'Options', uppercase: 'OPTIONS' },
  patch: { capitalized: 'Patch', uppercase: 'PATCH' },
  post: { capitalized: 'Post', uppercase: 'POST' },
  put: { capitalized: 'Put', uppercase: 'PUT' },
};

export const cachedSchemas: CachedSchemas = [
  { name: 'querystring', capitalized: 'Querystring' },
  { name: 'params', capitalized: 'Params' },
  { name: 'headers', capitalized: 'Headers' },
  { name: 'response', capitalized: 'Reply' },
];

export const cachedSchemasWithBody: CachedSchemas = [
  { name: 'body', capitalized: 'Body' },
  ...cachedSchemas,
];

export const methodsWithBody: Method[] = ['post', 'put', 'patch'];
