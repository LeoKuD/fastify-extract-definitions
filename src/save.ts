import { resolve } from 'path';
import { existsSync, promises } from 'fs';

export const save = async (text: string, path: string) => {
  const filePath = resolve(path);

  if (existsSync(filePath)) {
    const data = await promises.readFile(filePath);

    if (data.toString() === text) {
      return;
    }
  }

  await promises.writeFile(filePath, text);
};
