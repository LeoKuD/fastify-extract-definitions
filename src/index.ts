import { FastifyPluginCallback } from 'fastify';
import fastifyPlugin from 'fastify-plugin';
import fastifyRoutes from 'fastify-routes';

import { compile } from './compile';
import { prettify } from './prettify';
import { bannerComment, bannerImports } from './constants';
import { save } from './save';
import { ExtractorOptions } from './types';

export * from './types';

const plugin: FastifyPluginCallback<ExtractorOptions> = async (
  fastify,
  options = { enabled: false, outputs: {} },
  done
) => {
  if (!options.enabled) {
    done();
    return;
  }

  fastify.register(fastifyRoutes);

  fastify.addHook('onReady', async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const app = fastify as any;
    const definitions = app.getSchemas();

    const outputs = Object.keys(options.outputs);
    const baseComment = options.compilerOptions?.bannerComment || bannerComment;

    const [{ text, handlers }, prettier] = await Promise.all([
      compile(app.routes, definitions, options.compilerOptions),
      prettify(options.compilerOptions),
    ]);

    await Promise.all(
      outputs.map(async (output) => {
        const item = options.outputs[output];
        const isServer = item.target === 'serverTypes';

        const banner = baseComment + (isServer ? bannerImports : '');
        const beHandlers = isServer ? handlers : '';

        const formatted = await prettier(banner + text + beHandlers);
        await save(formatted, output);
      })
    );
  });

  done();
};

export default fastifyPlugin(plugin, {
  fastify: '>=3.x',
  name: 'fastify-extract-definitions',
});
