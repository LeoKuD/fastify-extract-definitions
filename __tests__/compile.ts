import { RouteOptions, HTTPMethods } from 'fastify';
import { JSONSchema4 } from 'json-schema';

import {
  compile,
  transformMethodLevel,
  transformResponse,
  transformRootLevel,
  transformSchemaLevel,
} from '../src/compile';
import { createSchemaObject } from '../src/utils';
import { cachedSchemasWithBody } from '../src/constants';
import { Route } from '../src/types';

const properties: Record<string, JSONSchema4> = {
  id: { type: 'string' },
  label: { type: 'string' },
};

const config = (route: string, method: HTTPMethods): RouteOptions => ({
  method,
  url: `${route}/${method}`,
  handler: console.log,
  schema: {
    body: createSchemaObject(undefined, {}),
    querystring: createSchemaObject(undefined, {}),
    params: createSchemaObject(undefined, {}),
    headers: createSchemaObject(undefined, {}),
    response: {
      200: createSchemaObject(undefined, {}),
    },
  },
});

const schema = (route: string): Route => [
  config(route, 'DELETE'),
  config(route, 'GET'),
  config(route, 'HEAD'),
  config(route, 'OPTIONS'),
  config(route, 'PATCH'),
  config(route, 'POST'),
  config(route, 'PUT'),
];

describe('Transform response', () => {
  test('Response without title transformed correctly', () => {
    const response = {
      200: createSchemaObject(undefined, properties),
    };

    expect(transformResponse('Test', response)).toEqual([
      createSchemaObject('TestStatus200', properties),
    ]);
  });

  test('Response with title transformed correctly', () => {
    const response = {
      200: createSchemaObject('Custom title', properties),
    };

    expect(transformResponse('Test', response)).toEqual([
      createSchemaObject('TestStatusCustomTitle', properties),
    ]);
  });

  test('Response without properties transformed correctly', () => {
    const response = {
      200: createSchemaObject(undefined, {}),
    };

    expect(transformResponse('Test', response)).toEqual([
      createSchemaObject('TestStatus200', {}),
    ]);
  });
});

describe('Transform schema level', () => {
  test('If no schema is specified, an empty object will be returned', () => {
    const schema = transformSchemaLevel(
      'CustomTitle',
      {},
      cachedSchemasWithBody
    );

    expect(schema).toEqual({});
  });

  test('Response transformed correctly', () => {
    const schema = transformSchemaLevel(
      'CustomTitle',
      {
        response: {
          200: createSchemaObject(undefined, properties),
          500: createSchemaObject('Custom error', {}),
        },
      },
      cachedSchemasWithBody
    );

    expect(schema).toEqual({
      Reply: {
        ...createSchemaObject('CustomTitleReply', undefined),
        oneOf: [
          createSchemaObject('CustomTitleReplyStatus200', properties),
          createSchemaObject('CustomTitleReplyStatusCustomError', {}),
        ],
      },
    });
  });

  test('Other fields transform correctly', () => {
    const schema = transformSchemaLevel(
      'MyTitle',
      {
        body: createSchemaObject('Custom title 1', properties),
        querystring: createSchemaObject('Custom title 2', properties),
        params: createSchemaObject('Custom title 3', properties),
        headers: createSchemaObject('Custom title 4', properties),
      },
      cachedSchemasWithBody
    );

    expect(schema).toEqual({
      Body: createSchemaObject('Custom title 1', properties),
      Headers: createSchemaObject('Custom title 4', properties),
      Params: createSchemaObject('Custom title 3', properties),
      Querystring: createSchemaObject('Custom title 2', properties),
    });
  });
});

describe('Transform method level', () => {
  test('If the config is empty, an empty object will be returned', () => {
    expect(transformMethodLevel('MyTitle', [], false)).toEqual({});
  });

  test('Body for all requests except POST, PUT and PATCH will be ignored', () => {
    const result = transformMethodLevel('MyTitle', schema(''), false);

    expect(result.DELETE.properties.Body).toBeUndefined();
    expect(result.GET.properties.Body).toBeUndefined();
    expect(result.HEAD.properties.Body).toBeUndefined();
    expect(result.OPTIONS.properties.Body).toBeUndefined();

    expect(result.PATCH.properties.Body).toBeDefined();
    expect(result.POST.properties.Body).toBeDefined();
    expect(result.PUT.properties.Body).toBeDefined();
  });

  const output = {
    DELETE: createSchemaObject('MyTitleDelete', {
      Querystring: createSchemaObject('MyTitleDeleteQuerystring', {}),
      Params: createSchemaObject('MyTitleDeleteParams', {}),
      Headers: createSchemaObject('MyTitleDeleteHeaders', {}),
      Reply: {
        ...createSchemaObject('MyTitleDeleteReply', undefined),
        oneOf: [createSchemaObject('MyTitleDeleteReplyStatus200', {})],
      },
    }),
    GET: createSchemaObject('MyTitleGet', {
      Querystring: createSchemaObject('MyTitleGetQuerystring', {}),
      Params: createSchemaObject('MyTitleGetParams', {}),
      Headers: createSchemaObject('MyTitleGetHeaders', {}),
      Reply: {
        ...createSchemaObject('MyTitleGetReply', undefined),
        oneOf: [createSchemaObject('MyTitleGetReplyStatus200', {})],
      },
    }),
    OPTIONS: createSchemaObject('MyTitleOptions', {
      Querystring: createSchemaObject('MyTitleOptionsQuerystring', {}),
      Params: createSchemaObject('MyTitleOptionsParams', {}),
      Headers: createSchemaObject('MyTitleOptionsHeaders', {}),
      Reply: {
        ...createSchemaObject('MyTitleOptionsReply', undefined),
        oneOf: [createSchemaObject('MyTitleOptionsReplyStatus200', {})],
      },
    }),
    PATCH: createSchemaObject('MyTitlePatch', {
      Body: createSchemaObject('MyTitlePatchBody', {}),
      Querystring: createSchemaObject('MyTitlePatchQuerystring', {}),
      Params: createSchemaObject('MyTitlePatchParams', {}),
      Headers: createSchemaObject('MyTitlePatchHeaders', {}),
      Reply: {
        ...createSchemaObject('MyTitlePatchReply', undefined),
        oneOf: [createSchemaObject('MyTitlePatchReplyStatus200', {})],
      },
    }),
    POST: createSchemaObject('MyTitlePost', {
      Body: createSchemaObject('MyTitlePostBody', {}),
      Querystring: createSchemaObject('MyTitlePostQuerystring', {}),
      Params: createSchemaObject('MyTitlePostParams', {}),
      Headers: createSchemaObject('MyTitlePostHeaders', {}),
      Reply: {
        ...createSchemaObject('MyTitlePostReply', undefined),
        oneOf: [createSchemaObject('MyTitlePostReplyStatus200', {})],
      },
    }),
    PUT: createSchemaObject('MyTitlePut', {
      Body: createSchemaObject('MyTitlePutBody', {}),
      Querystring: createSchemaObject('MyTitlePutQuerystring', {}),
      Params: createSchemaObject('MyTitlePutParams', {}),
      Headers: createSchemaObject('MyTitlePutHeaders', {}),
      Reply: {
        ...createSchemaObject('MyTitlePutReply', undefined),
        oneOf: [createSchemaObject('MyTitlePutReplyStatus200', {})],
      },
    }),
  };

  test('Transformer with HEAD ignoring working correctly', () => {
    expect(transformMethodLevel('MyTitle', schema(''), true)).toEqual(output);
  });

  test('Transformer without HEAD ignoring working correctly', () => {
    expect(transformMethodLevel('MyTitle', schema(''), false)).toEqual({
      ...output,
      HEAD: createSchemaObject('MyTitleHead', {
        Querystring: createSchemaObject('MyTitleHeadQuerystring', {}),
        Params: createSchemaObject('MyTitleHeadParams', {}),
        Headers: createSchemaObject('MyTitleHeadHeaders', {}),
        Reply: {
          ...createSchemaObject('MyTitleHeadReply', undefined),
          oneOf: [createSchemaObject('MyTitleHeadReplyStatus200', {})],
        },
      }),
    });
  });
});

describe('Transform root level', () => {
  test('Transformer works correctly', () => {
    const routes: [route: string, config: Route][] = [
      ['/foo', schema('/foo')],
      ['/bar', schema('/bar')],
    ];
    const result = transformRootLevel(routes, false);

    expect(result).toMatchSnapshot();
  });
});

describe('Compile', () => {
  const definitions = {
    enums: {
      $id: 'enums',
      type: 'object',
      title: 'enums',
      required: ['mode'],
      properties: {
        mode: {
          title: 'Mode',
          type: 'string',
          enum: ['production', 'staging', 'development', 'test'],
        },
      },
      additionalProperties: false,
    },
  };

  describe('Handle refs', () => {
    const routes = new Map<string, Route>();

    routes.set('/', [
      {
        url: '/',
        method: 'GET',
        handler: console.log,
        schema: {
          params: { $ref: 'enums#/properties/mode' },
          response: {
            200: { $ref: 'enums#/properties/mode' },
          },
        },
      },
    ]);

    test('Compiler working correctly', async () => {
      const result = await compile(routes, definitions, false);

      expect(result).toMatchSnapshot();
    });

    test('Compiler with `unreachableDefinitions` option working correctly', async () => {
      const result = await compile(routes, definitions, false, {
        unreachableDefinitions: true,
      });

      expect(result).toMatchSnapshot();
    });
  });

  describe('Handle duplicates', () => {
    const routes = new Map<string, Route>();

    routes.set('/a', [
      {
        url: '/a',
        method: 'GET',
        handler: console.log,
        schema: {
          params: { $ref: 'enums#/properties/mode' },
        },
      },
    ]);

    routes.set('/b', [
      {
        url: '/b',
        method: 'GET',
        handler: console.log,
        schema: {
          params: { title: 'Mode', type: 'number' },
        },
      },
    ]);

    routes.set('/c', [
      {
        url: '/c',
        method: 'GET',
        handler: console.log,
        schema: {
          params: { title: 'Mode', type: 'string' },
        },
      },
    ]);

    routes.set('/d', [
      {
        url: '/d',
        method: 'GET',
        handler: console.log,
        schema: {
          params: { title: 'Mode', type: 'string' },
        },
      },
    ]);

    test('Compiler working correctly', async () => {
      const result = await compile(routes, definitions, false);

      expect(result).toMatchSnapshot();
    });

    test('Compiler with `unreachableDefinitions` option working correctly', async () => {
      const result = await compile(routes, definitions, false, {
        unreachableDefinitions: true,
      });

      expect(result).toMatchSnapshot();
    });
  });
});
