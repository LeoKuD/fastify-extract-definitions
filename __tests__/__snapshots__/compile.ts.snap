// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compile Handle duplicates Compiler with \`unreachableDefinitions\` option working correctly 1`] = `
"
export type A1GetParams = (\\"production\\" | \\"staging\\" | \\"development\\" | \\"test\\")

export type Mode = number

export type Mode1 = string

export type Mode2 = string

export type Mode3 = (\\"production\\" | \\"staging\\" | \\"development\\" | \\"test\\")

export interface API {
\\"/a\\": {
GET: A1Get
}
\\"/b\\": {
GET: B1Get
}
\\"/c\\": {
GET: C1Get
}
\\"/d\\": {
GET: D1Get
}
}

export interface A1Get {
Querystring: unknown
Params: A1GetParams
Headers: unknown
Reply: unknown
}

export interface B1Get {
Querystring: unknown
Params: Mode
Headers: unknown
Reply: unknown
}

export interface C1Get {
Querystring: unknown
Params: Mode1
Headers: unknown
Reply: unknown
}

export interface D1Get {
Querystring: unknown
Params: Mode2
Headers: unknown
Reply: unknown
}

/**
 * This interface was referenced by \`API\`'s JSON-Schema
 * via the \`definition\` \\"enums\\".
 */
export interface Enums {
mode: Mode3
}
"
`;

exports[`Compile Handle duplicates Compiler working correctly 1`] = `
"
export type AGetParams = (\\"production\\" | \\"staging\\" | \\"development\\" | \\"test\\")

export type Mode = number

export type Mode1 = string

export type Mode2 = string

export interface API {
\\"/a\\": {
GET: AGet
}
\\"/b\\": {
GET: BGet
}
\\"/c\\": {
GET: CGet
}
\\"/d\\": {
GET: DGet
}
}

export interface AGet {
Querystring: unknown
Params: AGetParams
Headers: unknown
Reply: unknown
}

export interface BGet {
Querystring: unknown
Params: Mode
Headers: unknown
Reply: unknown
}

export interface CGet {
Querystring: unknown
Params: Mode1
Headers: unknown
Reply: unknown
}

export interface DGet {
Querystring: unknown
Params: Mode2
Headers: unknown
Reply: unknown
}
"
`;

exports[`Compile Handle refs Compiler with \`unreachableDefinitions\` option working correctly 1`] = `
"
export type RootGetParams = (\\"production\\" | \\"staging\\" | \\"development\\" | \\"test\\")

export type RootGetReply = Mode

export type Mode = (\\"production\\" | \\"staging\\" | \\"development\\" | \\"test\\")

export interface API {
\\"/\\": {
GET: RootGet
}
}

export interface RootGet {
Querystring: unknown
Params: RootGetParams
Headers: unknown
Reply: RootGetReply
}

/**
 * This interface was referenced by \`API\`'s JSON-Schema
 * via the \`definition\` \\"enums\\".
 */
export interface Enums {
mode: Mode
}
"
`;

exports[`Compile Handle refs Compiler working correctly 1`] = `
"
export type RootGetParams = (\\"production\\" | \\"staging\\" | \\"development\\" | \\"test\\")

export type RootGetReply = Mode

export type Mode = (\\"production\\" | \\"staging\\" | \\"development\\" | \\"test\\")

export interface API {
\\"/\\": {
GET: RootGet
}
}

export interface RootGet {
Querystring: unknown
Params: RootGetParams
Headers: unknown
Reply: RootGetReply
}
"
`;

exports[`Transform root level Transformer works correctly 1`] = `
Object {
  "/bar": Object {
    "additionalProperties": false,
    "properties": Object {
      "DELETE": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarDeleteHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarDeleteParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarDeleteQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "BarDeleteReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "BarDeleteReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "BarDelete",
        "type": "object",
      },
      "GET": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarGetHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarGetParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarGetQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "BarGetReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "BarGetReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "BarGet",
        "type": "object",
      },
      "HEAD": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarHeadHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarHeadParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarHeadQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "BarHeadReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "BarHeadReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "BarHead",
        "type": "object",
      },
      "OPTIONS": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarOptionsHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarOptionsParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarOptionsQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "BarOptionsReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "BarOptionsReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "BarOptions",
        "type": "object",
      },
      "PATCH": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Body": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPatchBody",
            "type": "object",
          },
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPatchHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPatchParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPatchQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "BarPatchReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "BarPatchReply",
            "type": "object",
          },
        },
        "required": Array [
          "Body",
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "BarPatch",
        "type": "object",
      },
      "POST": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Body": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPostBody",
            "type": "object",
          },
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPostHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPostParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPostQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "BarPostReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "BarPostReply",
            "type": "object",
          },
        },
        "required": Array [
          "Body",
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "BarPost",
        "type": "object",
      },
      "PUT": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Body": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPutBody",
            "type": "object",
          },
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPutHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPutParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "BarPutQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "BarPutReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "BarPutReply",
            "type": "object",
          },
        },
        "required": Array [
          "Body",
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "BarPut",
        "type": "object",
      },
    },
    "required": Array [
      "DELETE",
      "GET",
      "HEAD",
      "OPTIONS",
      "PATCH",
      "POST",
      "PUT",
    ],
    "title": undefined,
    "type": "object",
  },
  "/foo": Object {
    "additionalProperties": false,
    "properties": Object {
      "DELETE": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooDeleteHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooDeleteParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooDeleteQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "FooDeleteReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "FooDeleteReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "FooDelete",
        "type": "object",
      },
      "GET": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooGetHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooGetParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooGetQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "FooGetReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "FooGetReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "FooGet",
        "type": "object",
      },
      "HEAD": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooHeadHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooHeadParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooHeadQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "FooHeadReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "FooHeadReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "FooHead",
        "type": "object",
      },
      "OPTIONS": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooOptionsHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooOptionsParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooOptionsQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "FooOptionsReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "FooOptionsReply",
            "type": "object",
          },
        },
        "required": Array [
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "FooOptions",
        "type": "object",
      },
      "PATCH": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Body": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPatchBody",
            "type": "object",
          },
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPatchHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPatchParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPatchQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "FooPatchReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "FooPatchReply",
            "type": "object",
          },
        },
        "required": Array [
          "Body",
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "FooPatch",
        "type": "object",
      },
      "POST": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Body": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPostBody",
            "type": "object",
          },
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPostHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPostParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPostQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "FooPostReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "FooPostReply",
            "type": "object",
          },
        },
        "required": Array [
          "Body",
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "FooPost",
        "type": "object",
      },
      "PUT": Object {
        "additionalProperties": false,
        "description": undefined,
        "properties": Object {
          "Body": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPutBody",
            "type": "object",
          },
          "Headers": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPutHeaders",
            "type": "object",
          },
          "Params": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPutParams",
            "type": "object",
          },
          "Querystring": Object {
            "additionalProperties": false,
            "properties": Object {},
            "required": Array [],
            "title": "FooPutQuerystring",
            "type": "object",
          },
          "Reply": Object {
            "additionalProperties": false,
            "oneOf": Array [
              Object {
                "additionalProperties": false,
                "properties": Object {},
                "required": Array [],
                "title": "FooPutReplyStatus200",
                "type": "object",
              },
            ],
            "properties": undefined,
            "required": undefined,
            "title": "FooPutReply",
            "type": "object",
          },
        },
        "required": Array [
          "Body",
          "Querystring",
          "Params",
          "Headers",
          "Reply",
        ],
        "title": "FooPut",
        "type": "object",
      },
    },
    "required": Array [
      "DELETE",
      "GET",
      "HEAD",
      "OPTIONS",
      "PATCH",
      "POST",
      "PUT",
    ],
    "title": undefined,
    "type": "object",
  },
}
`;
